### PR TITLE
[openshift-apiserver-4.16-kubernetes-1.29.2] Add carry patches from 4.15

### PR DESCRIPTION
**Pull request by liouk**
This PR adds carried commits from `openshift-apiserver-4.15-kubernetes-1.28.3`.

Proof PRs:
- [oauth-server#142](https://github.com/openshift/oauth-server/pull/142)
- [oauth-apiserver#108](https://github.com/openshift/oauth-apiserver/pull/108)
- [openshift-apiserver#421](https://github.com/openshift/openshift-apiserver/pull/421)

### CODE REVIEW

File: pkg/server/config.go [hunk 1]
```diff
@@ -730,6 +762,39 @@ var allowedMediaTypes = []string{
 	runtime.ContentTypeProtobuf,
 }
 
+func eventReference() (*corev1.ObjectReference, error) {
```
Comment (liouk) [inline pkg/server/config.go:765]: Merge conflict resolved here by keeping both `allowedMediaTypes` and `eventReference()` definitions.

File: pkg/storage/storagebackend/factory/etcd3.go [hunk 2]
```diff
@@ -454,7 +455,7 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc fu
 	if transformer == nil {
 		transformer = identity.NewEncryptCheckTransformer()
 	}
-	return etcd3.New(client, c.Codec, newFunc, newListFunc, c.Prefix, resourcePrefix, c.GroupResource, transformer, c.LeaseManagerConfig), destroyFunc, nil
+	return etcd3retry.NewRetryingEtcdStorage(etcd3.New(client, c.Codec, newFunc, newListFunc, c.Prefix, resourcePrefix, c.GroupResource, transformer, c.LeaseManagerConfig)), destroyFunc, nil
```
Comment (liouk) [inline pkg/storage/storagebackend/factory/etcd3.go:458]: Merge conflict resolved here by wrapping `etcd3.New()` return value into `etcd3retry.NewRetryingEtcdStorage()`.

**Comment by p0lyn0mial** [comment]
I've compared this PR against https://github.com/openshift/kubernetes-apiserver/commits/openshift-apiserver-4.15-kubernetes-1.28.3/

`[UPSTREAM: <carry>: Enable UnauthenticatedHTTP2DOSMitigation by default]` was dropped because the `UnauthenticatedHTTP2DOSMitigation` is `enabled` by default in 1.29 (https://github.com/openshift/kubernetes-apiserver/blob/openshift-apiserver-4.16-kubernetes-1.29.2/pkg/features/kube_features.go#L318)

`[UPSTREAM: 118886: KEP-1040: Deep disablement for APF based on --enable-priority-and-fairness.](https://github.com/openshift/kubernetes-apiserver/commit/b4c0eac7a9f1537a46f5b5339ce3e8191d34414c)` was dropped because the commit is part of 1.29 release.

/lgtm

**Comment by p0lyn0mial** [comment]
I have checked https://github.com/openshift/oauth-apiserver/pull/108, in general it looks OK. I think we can merge this PR.

**Pull request merged**